### PR TITLE
deep-freeze: Updated to use Mapped Types and align with Object.freeze in lib.d.ts

### PR DESCRIPTION
**Pull request by aluanhaddad**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/substack/deep-freeze/blob/master/index.js
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

**Remarks:**

Updated the declaration to take advantage of Mapped Types. Specifically, I added a `DeepReadonly` recursive generic mapped type. 

Additionally, I took the opportunity to add overloads that mirroring those of `Object.freeze` as specified in the latest `lib.d.ts`.

I removed the interface as it is now unused.

### CODE REVIEW

File: types/deep-freeze/index.d.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-declare namespace DeepFreeze {
-   export interface DeepFreezeInterface {
-      <T>(obj: T): T;
-    }
-}
+declare function deepFreeze<T>(a: T[]): ReadonlyArray<DeepReadonly<T>>;
+declare function deepFreeze<T extends Function>(f: T): T;
+declare function deepFreeze<T>(o: T): DeepReadonly<T>;
+
+type DeepReadonly<T> = Readonly<{ [P in keyof T]: DeepReadonly<T[P]> }>;
```
Comment (DanielRosenwasser) [inline types/deep-freeze/index.d.ts]: One problem is that consumers can no longer reference `DeepReadonly`, so ideally you'd do function/namespace merging, export `DeepReadonly` from the namespace, and `export =` the declarations. So


```ts
declare function deepFreeze<T>(a: T[]): ReadonlyArray<deepFreeze.DeepReadonly<T>>;
declare function deepFreeze<T extends Function>(f: T): T;
declare function deepFreeze<T>(o: T): deepFreeze.DeepReadonly<T>;

declare namespace deepFreeze {
    export type DeepReadonly<T> = Readonly<{ [P in keyof T]: DeepReadonly<T[P]> }>;
}

export deepFreeze
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/deep-freeze/deep-freeze-tests.ts [hunk 2]
```diff
@@ -1,8 +1,7 @@
...
-
 import df = require('deep-freeze');
 
 class Foo {
-	foo: string;
+    foo: string;
 }
-var foo:Foo = df(new Foo());
+const foo = df(new Foo());
+const items = [{id: 0, name: 'first'}];
```
Comment (DanielRosenwasser) [inline types/deep-freeze/deep-freeze-tests.ts]: Did you mean to call `df` on the value of `items`?

Comment (aluanhaddad) [inline types/deep-freeze/deep-freeze-tests.ts]: Yes... 😄

**Comment by DanielRosenwasser** [comment]
Sorry about the delay!

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**